Add Support for Oracle Autonomous DB

recursivecodes opened the pull request:
This PR adds support for Oracle Autonomous DB. Has been tested on MacOS (Catalina) and Linux. Requires Oracle Instant Client to be installed prior to running (like most Oracle tools). I will follow-up with a blog post on utilization after PR is merged.

tyleradams commented on json-sql:
```diff
@@ -35,6 +47,7 @@ def make_table(data, columns):
 
 class DB:
     connection = None
+    db_type = None
```
This isn't being used, can we remove it?

tyleradams commented on json-sql:
```diff
@@ -53,8 +66,10 @@ class DB:
         columns = self._column_names(table_name)
-
-        data = self.execute_query("select * from {};".format(table_name))
+        if self.__class__.__name__ == 'ORACLEADB':
```
Can we change this to override the read_table method in the ORACLEDB class instead of putting an if statement here?

tyleradams commented on json-sql:
```diff
@@ -168,8 +203,14 @@ Example usages:
+
     Insert data into a mysql DB:
-        echo ["insert into table (a,b,c) values(1, 2, 3);"] | json-sql query mysql user password localhost 3306 db
+        echo '["insert into table (a,b,c) values(1, 2, 3);"]' | json-sql query mysql user password localhost 3306 db
```
Thanks!

tyleradams commented on json-sql:
```diff
@@ -1,17 +1,29 @@
 
+# to avoid "TypeError: Object of type datetime is not JSON serializable"
+# per https://stackoverflow.com/a/35870294
+def datetime_handler(x):
```
Nice catch on the datetime objects.
We can simplify the code by just using the logic in jsonable that you implemented on line 25.

tyleradams commented on json-sql:
```diff
@@ -195,7 +236,7 @@ def main():
     elif method == "query":
         queries = json.load(sys.stdin)
-        print(json.dumps([db.execute_query(query) for query in queries]))
+        print(json.dumps([db.execute_query(query) for query in queries], default=datetime_handler))
```
Nice usage of default.

If we set default=jsonable, then serializing other types can be done by adding a method to jsonable.

tyleradams commented on json-sql:
```diff
@@ -1,17 +1,29 @@
 
 import pymysql
 import psycopg2
+import cx_Oracle
```
Similarly, can you please alphabetize this list as well?

tyleradams commented on json-sql:
```diff
@@ -84,6 +99,24 @@ class DB:
         return NotImplementedError
 
+class ORACLEADB(DB):
+    schema = None
```
It might be good style to declare class variables in python, I'm not sure, but since no other classes do this, if we remove this, we can keep the code more consistent.

recursivecodes replied:
I don't use python much - just my Java developer coming through! :)

tyleradams commented on json-sql:
```diff
@@ -84,6 +99,24 @@ class DB:
+class ORACLEADB(DB):
+    schema = None
+    def __init__(self, user, password, schema, tnsname, lib_dir=None):
+        if os.environ['TNS_ADMIN'] == "":
```
I'm not familiar with OracleDB, why does TNS_ADMIN need to be set? Although we check for it, I can't find where TNS_ADMIN is used.

recursivecodes replied:
`TNS_ADMIN` is required by the cx_Oracle library. It's the path to a "wallet" directory containing keys that are used to secure the connection to the Oracle DB. The library would throw an error if it's not set, but it's a little vague so I wanted to be more explicit.

tyleradams replied:
Great thanks, it's not blocking but if you'd like, you can add a comment here mentioning as such so the next reader isn't confused.

tyleradams commented on json-sql:
```diff
@@ -1,17 +1,29 @@
 #!/usr/bin/env python3
 
 import decimal
+import datetime
```
Personally, I'm a big fan of keeping lists like this alphabetical, can you please move datetime up to the top of the list?

tyleradams commented on requirements.txt:
```diff
@@ -6,3 +6,4 @@ PyMySQL==0.10.1
 PyYAML==5.3.1
 unidiff==0.6.0
 xmltodict==0.12.0
+cx_Oracle==8.1.0
```
Can you please alphabetize this file?

tyleradams reviewed (changes requested):
Great work, thanks for doing this!

I just have some small touchup requests to keep the code clean and one question about something I don't understand.

recursivecodes commented:
I think I've addressed all of your concerns in the latest commit (https://github.com/tyleradams/json-toolkit/pull/2/commits/e5da4f568063685f638cae50efef27f5cdf1d563).  Please forgive formatting/style errors as I don't work much with Python 😄.

recursivecodes commented:
Oh, hang on - missed a few comments...

recursivecodes commented:
More changes added in https://github.com/tyleradams/json-toolkit/pull/2/commits/c4702fa94c1ce7ccc12e9bd49ec0765afa9861b5. I'm not sure if I properly alphabetized the imports to your liking - let me know if not. It seemed like there were two groups of imports - both sorted individually. Guessing maybe one was for internal and the other for external deps?  Please let me know if I did this wrong! :)

tyleradams reviewed (approved):
LGTM

recursivecodes commented:
Comment added! https://github.com/tyleradams/json-toolkit/pull/2/commits/a9b9c1574db2cb42ecda0eadc7432c31bbfa7680
